# Customer Churn Analysis

## Project Objective

This project analyzes customer churn behavior using the Telco Customer Churn dataset.

The goal is to identify which customer characteristics are most associated with churn and later build a simple predictive model to estimate churn risk.

The main business question is:

**Which customers are most likely to churn, and what factors are linked with higher churn risk?**

This project uses Python for data validation, cleaning, exploratory analysis, and modeling. A cleaned dataset will later be exported for Power BI dashboarding.

In [1]:
from google.colab import files

uploaded = files.upload()

Saving Telco-Customer-Churn.csv to Telco-Customer-Churn.csv


## Project Setup

In this section, I import the required Python libraries and load the raw customer churn dataset into a pandas DataFrame.

The dataset contains customer level information such as tenure, contract type, monthly charges, internet services, payment method, and churn status.

In [2]:
import pandas as pd

df = pd.read_csv("Telco-Customer-Churn.csv")
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


## Phase 1: Data Validation and Cleaning

Before analyzing churn behavior, I first validated the dataset to make sure it was clean and reliable.

This phase checks:

- Dataset shape
- Column names
- Missing values
- Churn distribution
- Duplicate customer IDs
- Numeric ranges
- Data type issues

The key cleaning issue was the `TotalCharges` column. Although it represents a numeric value, it was initially stored as text. After converting it to numeric, 11 missing values appeared. These rows had `tenure = 0`, meaning the customers had not accumulated total charges yet. Therefore, the missing `TotalCharges` values were filled with 0.

In [3]:
df.shape

(7043, 21)

In [4]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [5]:
df.isnull().sum()

,0
customerID,0
gender,0
SeniorCitizen,0
Partner,0
Dependents,0
tenure,0
PhoneService,0
MultipleLines,0
InternetService,0
OnlineSecurity,0


In [6]:
df["Churn"].value_counts()

,count
Churn,
No,5174
Yes,1869


In [7]:
df["Churn"].value_counts(normalize=True) * 100

,proportion
Churn,
No,73.463013
Yes,26.536987


In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [9]:
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors="coerce")

In [10]:
df["TotalCharges"].isnull().sum()

np.int64(11)

In [11]:
df[df["TotalCharges"].isnull()]

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
488,4472-LVYGI,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,Yes,No,Two year,Yes,Bank transfer (automatic),52.55,NaN,No
753,3115-CZMZD,Male,0,No,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.25,NaN,No
936,5709-LVOEQ,Female,0,Yes,Yes,0,Yes,No,DSL,Yes,...,Yes,No,Yes,Yes,Two year,No,Mailed check,80.85,NaN,No
1082,4367-NUYAO,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.75,NaN,No
1340,1371-DWPAZ,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,Yes,No,Two year,No,Credit card (automatic),56.05,NaN,No
3331,7644-OMVMY,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,19.85,NaN,No
3826,3213-VVOLG,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.35,NaN,No
4380,2520-SGTTA,Female,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.00,NaN,No
5218,2923-ARZLG,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,One year,Yes,Mailed check,19.70,NaN,No
6670,4075-WKNIU,Female,0,Yes,Yes,0,Yes,Yes,DSL,No,...,Yes,Yes,Yes,No,Two year,No,Mailed check,73.35,NaN,No


In [12]:
df[df["TotalCharges"].isnull()][["customerID", "tenure", "MonthlyCharges", "TotalCharges", "Churn"]]

,customerID,tenure,MonthlyCharges,TotalCharges,Churn
488,4472-LVYGI,0,52.55,NaN,No
753,3115-CZMZD,0,20.25,NaN,No
936,5709-LVOEQ,0,80.85,NaN,No
1082,4367-NUYAO,0,25.75,NaN,No
1340,1371-DWPAZ,0,56.05,NaN,No
3331,7644-OMVMY,0,19.85,NaN,No
3826,3213-VVOLG,0,25.35,NaN,No
4380,2520-SGTTA,0,20.00,NaN,No
5218,2923-ARZLG,0,19.70,NaN,No
6670,4075-WKNIU,0,73.35,NaN,No


In [13]:
df["TotalCharges"] = df["TotalCharges"].fillna(0)

In [14]:
df.isnull().sum()

,0
customerID,0
gender,0
SeniorCitizen,0
Partner,0
Dependents,0
tenure,0
PhoneService,0
MultipleLines,0
InternetService,0
OnlineSecurity,0


In [15]:
df["customerID"].duplicated().sum()

np.int64(0)

In [16]:
df[["tenure", "MonthlyCharges", "TotalCharges"]].describe()

,tenure,MonthlyCharges,TotalCharges
count,7043.000000,7043.000000,7043.000000
mean,32.371149,64.761692,2279.734304
std,24.559481,30.090047,2266.794470
min,0.000000,18.250000,0.000000
25%,9.000000,35.500000,398.550000
50%,29.000000,70.350000,1394.550000
75%,55.000000,89.850000,3786.600000
max,72.000000,118.750000,8684.800000


In [17]:
df.to_csv("telco_churn_cleaned.csv", index=False)

In [18]:
from google.colab import files
files.download("telco_churn_cleaned.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Phase 1: Data Validation and Cleaning

- Dataset contains 7,043 rows and 21 columns.
- Target column is `Churn`.
- Churn distribution: 1,869 churned customers and 5,174 non-churned customers.
- Churn rate is 26.5%.
- `TotalCharges` was originally stored as text and converted to numeric.
- 11 missing `TotalCharges` values appeared after conversion.
- These rows had `tenure = 0`, so missing `TotalCharges` values were filled with 0.
- No duplicate customer IDs were found.

## Phase 2: Exploratory Churn Analysis

In this phase, I analyze churn rate across key customer groups to identify which customer characteristics are most associated with churn.

The focus areas are:
- Contract type
- Tenure group
- Monthly charge group
- Internet service
- Payment method
- Support and security services

In [19]:
df["ChurnFlag"] = df["Churn"].map({"Yes": 1, "No": 0})
df[["Churn", "ChurnFlag"]].head()

,Churn,ChurnFlag
0,No,0
1,No,0
2,Yes,1
3,No,0
4,Yes,1


In [20]:
overall_churn_rate = df["ChurnFlag"].mean() * 100
overall_churn_rate

np.float64(26.536987079369588)

In [21]:
churn_by_contract = (
    df.groupby("Contract")
      .agg(
          customers=("customerID", "count"),
          churned=("ChurnFlag", "sum"),
          churn_rate=("ChurnFlag", "mean")
      )
      .reset_index()
)

churn_by_contract["churn_rate"] = churn_by_contract["churn_rate"] * 100
churn_by_contract.sort_values("churn_rate", ascending=False)

,Contract,customers,churned,churn_rate
0,Month-to-month,3875,1655,42.709677
1,One year,1473,166,11.269518
2,Two year,1695,48,2.831858


In [22]:
df["TenureGroup"] = pd.cut(
    df["tenure"],
    bins=[-1, 12, 24, 48, 72],
    labels=["0-12 months", "13-24 months", "25-48 months", "49-72 months"]
)

churn_by_tenure = (
    df.groupby("TenureGroup", observed=False)
      .agg(
          customers=("customerID", "count"),
          churned=("ChurnFlag", "sum"),
          churn_rate=("ChurnFlag", "mean")
      )
      .reset_index()
)

churn_by_tenure["churn_rate"] = churn_by_tenure["churn_rate"] * 100
churn_by_tenure

,TenureGroup,customers,churned,churn_rate
0,0-12 months,2186,1037,47.438243
1,13-24 months,1024,294,28.710938
2,25-48 months,1594,325,20.388959
3,49-72 months,2239,213,9.513176


In [23]:
df["MonthlyChargeGroup"] = pd.cut(
    df["MonthlyCharges"],
    bins=[0, 35, 70, 90, 120],
    labels=["Low", "Medium", "High", "Very High"]
)

churn_by_charges = (
    df.groupby("MonthlyChargeGroup", observed=False)
      .agg(
          customers=("customerID", "count"),
          churned=("ChurnFlag", "sum"),
          churn_rate=("ChurnFlag", "mean")
      )
      .reset_index()
)

churn_by_charges["churn_rate"] = churn_by_charges["churn_rate"] * 100
churn_by_charges

,MonthlyChargeGroup,customers,churned,churn_rate
0,Low,1735,189,10.893372
1,Medium,1725,413,23.942029
2,High,1844,697,37.798265
3,Very High,1739,570,32.777458


In [24]:
churn_by_internet = (
    df.groupby("InternetService")
      .agg(
          customers=("customerID", "count"),
          churned=("ChurnFlag", "sum"),
          churn_rate=("ChurnFlag", "mean")
      )
      .reset_index()
)

churn_by_internet["churn_rate"] = churn_by_internet["churn_rate"] * 100
churn_by_internet.sort_values("churn_rate", ascending=False)

,InternetService,customers,churned,churn_rate
1,Fiber optic,3096,1297,41.892765
0,DSL,2421,459,18.959108
2,No,1526,113,7.404980


In [25]:
churn_by_payment = (
    df.groupby("PaymentMethod")
      .agg(
          customers=("customerID", "count"),
          churned=("ChurnFlag", "sum"),
          churn_rate=("ChurnFlag", "mean")
      )
      .reset_index()
)

churn_by_payment["churn_rate"] = churn_by_payment["churn_rate"] * 100
churn_by_payment.sort_values("churn_rate", ascending=False)

,PaymentMethod,customers,churned,churn_rate
2,Electronic check,2365,1071,45.285412
3,Mailed check,1612,308,19.106700
0,Bank transfer (automatic),1544,258,16.709845
1,Credit card (automatic),1522,232,15.243101


### Phase 2 Summary

Exploratory churn analysis showed several clear churn risk patterns:

- Contract type was one of the strongest churn indicators. Month to month customers had a churn rate of 42.71%, compared with 11.27% for one year contracts and 2.83% for two year contracts.
- Churn was heavily concentrated among newer customers. Customers with 0–12 months of tenure had a churn rate of 47.44%, while customers with 49–72 months of tenure churned at only 9.51%.
- Monthly charges were associated with churn risk. The high monthly charge group had the highest churn rate at 37.80%, although churn did not increase perfectly linearly across all charge groups.
- Fiber optic customers had a churn rate of 41.89%, much higher than DSL customers and customers with no internet service.
- Customers using electronic check had the highest churn rate by payment method at 45.29%.

Overall, churn appears to be most concentrated among month to month customers, early tenure customers, fiber optic users, high monthly charge customers, and electronic check users.

## Phase 3: Feature Engineering

In this phase, I create new features based on the churn patterns identified during exploratory analysis.

The goal is to convert important churn signals into clear flags and groups that can be used for modeling and dashboarding.

Key engineered features include:

- ChurnFlag
- TenureGroup
- MonthlyChargeGroup
- HighRiskContract
- EarlyTenureFlag
- ElectronicCheckFlag
- FiberOpticFlag

In [26]:
df["ChurnFlag"] = df["Churn"].map({"Yes": 1, "No": 0})

df[["Churn", "ChurnFlag"]].head()

,Churn,ChurnFlag
0,No,0
1,No,0
2,Yes,1
3,No,0
4,Yes,1


In [27]:
df["TenureGroup"] = pd.cut(
    df["tenure"],
    bins=[-1, 12, 24, 48, 72],
    labels=["0-12 months", "13-24 months", "25-48 months", "49-72 months"]
)

df[["tenure", "TenureGroup"]].head()

,tenure,TenureGroup
0,1,0-12 months
1,34,25-48 months
2,2,0-12 months
3,45,25-48 months
4,2,0-12 months


In [28]:
df["MonthlyChargeGroup"] = pd.cut(
    df["MonthlyCharges"],
    bins=[0, 35, 70, 90, 120],
    labels=["Low", "Medium", "High", "Very High"]
)

df[["MonthlyCharges", "MonthlyChargeGroup"]].head()

,MonthlyCharges,MonthlyChargeGroup
0,29.85,Low
1,56.95,Medium
2,53.85,Medium
3,42.30,Medium
4,70.70,High


In [29]:
df["HighRiskContract"] = df["Contract"].apply(
    lambda x: 1 if x == "Month-to-month" else 0
)

df[["Contract", "HighRiskContract"]].head()

,Contract,HighRiskContract
0,Month-to-month,1
1,One year,0
2,Month-to-month,1
3,One year,0
4,Month-to-month,1


In [30]:
df["EarlyTenureFlag"] = df["tenure"].apply(
    lambda x: 1 if x <= 12 else 0
)

df[["tenure", "EarlyTenureFlag"]].head()

,tenure,EarlyTenureFlag
0,1,1
1,34,0
2,2,1
3,45,0
4,2,1


In [31]:
df["ElectronicCheckFlag"] = df["PaymentMethod"].apply(
    lambda x: 1 if x == "Electronic check" else 0
)

df[["PaymentMethod", "ElectronicCheckFlag"]].head()

,PaymentMethod,ElectronicCheckFlag
0,Electronic check,1
1,Mailed check,0
2,Mailed check,0
3,Bank transfer (automatic),0
4,Electronic check,1


In [32]:
df["FiberOpticFlag"] = df["InternetService"].apply(
    lambda x: 1 if x == "Fiber optic" else 0
)

df[["InternetService", "FiberOpticFlag"]].head()

,InternetService,FiberOpticFlag
0,DSL,0
1,DSL,0
2,DSL,0
3,DSL,0
4,Fiber optic,1


In [33]:
df[
    [
        "ChurnFlag",
        "TenureGroup",
        "MonthlyChargeGroup",
        "HighRiskContract",
        "EarlyTenureFlag",
        "ElectronicCheckFlag",
        "FiberOpticFlag"
    ]
].head()

,ChurnFlag,TenureGroup,MonthlyChargeGroup,HighRiskContract,EarlyTenureFlag,ElectronicCheckFlag,FiberOpticFlag
0,0,0-12 months,Low,1,1,1,0
1,0,25-48 months,Medium,0,0,0,0
2,1,0-12 months,Medium,1,1,0,0
3,0,25-48 months,Medium,0,0,0,0
4,1,0-12 months,High,1,1,1,1


In [34]:
df[
    [
        "HighRiskContract",
        "EarlyTenureFlag",
        "ElectronicCheckFlag",
        "FiberOpticFlag"
    ]
].sum()

,0
HighRiskContract,3875
EarlyTenureFlag,2186
ElectronicCheckFlag,2365
FiberOpticFlag,3096


In [35]:
df["RiskFactorCount"] = (
    df["HighRiskContract"]
    + df["EarlyTenureFlag"]
    + df["ElectronicCheckFlag"]
    + df["FiberOpticFlag"]
)

df[["HighRiskContract", "EarlyTenureFlag", "ElectronicCheckFlag", "FiberOpticFlag", "RiskFactorCount"]].head()

,HighRiskContract,EarlyTenureFlag,ElectronicCheckFlag,FiberOpticFlag,RiskFactorCount
0,1,1,1,0,3
1,0,0,0,0,0
2,1,1,0,0,2
3,0,0,0,0,0
4,1,1,1,1,4


In [36]:
churn_by_risk_count = (
    df.groupby("RiskFactorCount")
      .agg(
          customers=("customerID", "count"),
          churned=("ChurnFlag", "sum"),
          churn_rate=("ChurnFlag", "mean")
      )
      .reset_index()
)

churn_by_risk_count["churn_rate"] = churn_by_risk_count["churn_rate"] * 100

churn_by_risk_count

,RiskFactorCount,customers,churned,churn_rate
0,0,1808,54,2.986726
1,1,1498,156,10.413885
2,2,1818,522,28.712871
3,3,1288,688,53.416149
4,4,631,449,71.156894


### Phase 3 Summary

Feature engineering converted the strongest churn signals from exploratory analysis into clear risk flags.

The four main churn risk indicators were:

- Month to month contract
- Tenure of 12 months or less
- Electronic check payment method
- Fiber optic internet service

A combined `RiskFactorCount` was created by adding these four flags together.

Churn increased sharply as the number of risk factors increased:

| Risk Factor Count | Customers | Churn Rate |
|---:|---:|---:|
| 0 | 1,808 | 2.99% |
| 1 | 1,498 | 10.41% |
| 2 | 1,818 | 28.71% |
| 3 | 1,288 | 53.42% |
| 4 | 631 | 71.16% |

Customers with no major risk factors had a churn rate of only 2.99%, while customers with all four risk factors had a churn rate of 71.16%.

This suggests that churn risk compounds strongly when multiple high-risk customer characteristics appear together.

## Phase 4: Churn Modeling

In this phase, I build a simple predictive model to estimate customer churn risk.

The goal is not only to predict churn, but also to understand which customer characteristics are most useful for identifying high risk customers.

A logistic regression model is used because it is simple, explainable, and suitable for binary classification problems like churn prediction.

In [37]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, confusion_matrix, classification_report

In [39]:
features = [
    "tenure",
    "MonthlyCharges",
    "TotalCharges",
    "HighRiskContract",
    "EarlyTenureFlag",
    "ElectronicCheckFlag",
    "FiberOpticFlag",
    "RiskFactorCount"
]

X = df[features]
y = df["ChurnFlag"]

In [40]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [41]:
model = LogisticRegression(max_iter=1000)

model.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

In [42]:
y_pred = model.predict(X_test)

In [43]:
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)

accuracy, precision, recall

(0.7863733144073811, 0.6280701754385964, 0.4786096256684492)

In [44]:
confusion_matrix(y_test, y_pred)

array([[929, 106],
       [195, 179]])

In [45]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.83      0.90      0.86      1035
           1       0.63      0.48      0.54       374

    accuracy                           0.79      1409
   macro avg       0.73      0.69      0.70      1409
weighted avg       0.77      0.79      0.78      1409



In [46]:
coefficients = pd.DataFrame({
    "feature": features,
    "coefficient": model.coef_[0]
})

coefficients["abs_coefficient"] = coefficients["coefficient"].abs()

coefficients.sort_values("abs_coefficient", ascending=False)

,feature,coefficient,abs_coefficient
7,RiskFactorCount,0.595101,0.595101
3,HighRiskContract,0.577361,0.577361
6,FiberOpticFlag,0.157843,0.157843
4,EarlyTenureFlag,-0.129355,0.129355
0,tenure,-0.029818,0.029818
1,MonthlyCharges,0.013469,0.013469
5,ElectronicCheckFlag,-0.010748,0.010748
2,TotalCharges,0.000064,0.000064


### Phase 4 Summary

A logistic regression model was built to predict customer churn using tenure, charges, and engineered churn risk flags.

The model was evaluated using accuracy, precision, and recall.

| Metric | Value |
|---|---:|
| Accuracy | X |
| Precision | X |
| Recall | X |

For churn prediction, recall is especially important because it measures how many actual churned customers were correctly identified.

The model also showed which features were most influential in predicting churn based on logistic regression coefficients.

In [47]:
features_model_a = [
    "tenure",
    "MonthlyCharges",
    "TotalCharges",
    "HighRiskContract",
    "EarlyTenureFlag",
    "ElectronicCheckFlag",
    "FiberOpticFlag"
]

X = df[features_model_a]
y = df["ChurnFlag"]

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

model_a = LogisticRegression(max_iter=1000)
model_a.fit(X_train, y_train)

y_pred_a = model_a.predict(X_test)

accuracy_a = accuracy_score(y_test, y_pred_a)
precision_a = precision_score(y_test, y_pred_a)
recall_a = recall_score(y_test, y_pred_a)

accuracy_a, precision_a, recall_a

(0.7856635911994322, 0.6258741258741258, 0.4786096256684492)

In [48]:
coefficients_a = pd.DataFrame({
    "feature": features_model_a,
    "coefficient": model_a.coef_[0]
})

coefficients_a["abs_coefficient"] = coefficients_a["coefficient"].abs()

coefficients_a.sort_values("abs_coefficient", ascending=False)

,feature,coefficient,abs_coefficient
3,HighRiskContract,1.168522,1.168522
6,FiberOpticFlag,0.743682,0.743682
5,ElectronicCheckFlag,0.581348,0.581348
4,EarlyTenureFlag,0.456134,0.456134
0,tenure,-0.030160,0.030160
1,MonthlyCharges,0.013606,0.013606
2,TotalCharges,0.000065,0.000065


In [51]:
confusion_matrix(y_test, y_pred_a)

array([[928, 107],
       [195, 179]])

In [49]:
features_model_b = [
    "tenure",
    "MonthlyCharges",
    "TotalCharges",
    "RiskFactorCount"
]

X = df[features_model_b]
y = df["ChurnFlag"]

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

model_b = LogisticRegression(max_iter=1000)
model_b.fit(X_train, y_train)

y_pred_b = model_b.predict(X_test)

accuracy_b = accuracy_score(y_test, y_pred_b)
precision_b = precision_score(y_test, y_pred_b)
recall_b = recall_score(y_test, y_pred_b)

accuracy_b, precision_b, recall_b

(0.7799858055358411, 0.6142857142857143, 0.45989304812834225)

In [50]:
coefficients_b = pd.DataFrame({
    "feature": features_model_b,
    "coefficient": model_b.coef_[0]
})

coefficients_b["abs_coefficient"] = coefficients_b["coefficient"].abs()

coefficients_b.sort_values("abs_coefficient", ascending=False)

,feature,coefficient,abs_coefficient
3,RiskFactorCount,0.741111,0.741111
0,tenure,-0.025163,0.025163
1,MonthlyCharges,0.015349,0.015349
2,TotalCharges,0.000008,0.000008


In [52]:
confusion_matrix(y_test, y_pred_b)

array([[927, 108],
       [202, 172]])

### Phase 4 Summary

Two logistic regression models were tested.

**Model A** used individual churn-risk flags:
- Month to month contract
- Early tenure
- Electronic check payment method
- Fiber optic internet service
- Tenure
- Monthly charges
- Total charges

**Model B** used a combined `RiskFactorCount` instead of individual flags.

Model A performed slightly better and was more explainable, so it was selected as the main baseline churn model.

| Metric | Model A | Model B |
|---|---:|---:|
| Accuracy | 78.57% | 78.00% |
| Precision | 62.59% | 61.43% |
| Recall | 47.86% | 45.99% |

Model A achieved 78.57% accuracy and 62.59% precision, but recall was only 47.86%. This means the model correctly identified some churned customers but still missed a large portion of actual churners.

The most influential churn predictors in Model A were:

| Feature | Coefficient |
|---|---:|
| Month to month contract | 1.17 |
| Fiber optic internet | 0.74 |
| Electronic check payment | 0.58 |
| Early tenure | 0.46 |
| Tenure | -0.03 |

The positive coefficients indicate higher churn likelihood, while the negative tenure coefficient suggests that longer tenure customers are less likely to churn.

Overall, the model provides a useful baseline for understanding churn risk, but recall would need improvement before using it for proactive retention targeting.

## Phase 5: Export Cleaned Dataset for Power BI

In this phase, I export the cleaned and feature engineered dataset for dashboarding in Power BI.

The exported dataset includes the original customer attributes along with engineered churn risk fields such as:

- ChurnFlag
- TenureGroup
- MonthlyChargeGroup
- HighRiskContract
- EarlyTenureFlag
- ElectronicCheckFlag
- FiberOpticFlag
- RiskFactorCount

These fields will be used to build a churn analysis dashboard focused on customer risk segmentation and retention recommendations.

In [53]:
df.to_csv("telco_churn_cleaned_model_ready.csv", index=False)

In [54]:
from google.colab import files
files.download("telco_churn_cleaned_model_ready.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>